internal/configfsm/fsm.go

```diff
@@ -0,0 +1,315 @@
+package configfsm
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/qmuntal/stateless"
+)
+
+type Trigger string
+
+const (
+	// Transitions to Missing state
+	StartToMissing           Trigger = "Start->Missing"
+	ClusterNotReadyToMissing Trigger = "ClusterNotReady->Missing"
+	InProgressToMissing      Trigger = "InProgress->Missing"
+	CompletedToMissing       Trigger = "Completed->Missing"
+	TimedOutToMissing        Trigger = "TimedOut->Missing"
+	OutOfDateToMissing       Trigger = "OutOfDate->Missing"
+
+	// Transitions to ClusterNotReady state
+	MissingToClusterNotReady    Trigger = "Missing->ClusterNotReady"
+	InProgressToClusterNotReady Trigger = "InProgress->ClusterNotReady"
+	TimedOutToClusterNotReady   Trigger = "TimedOut->ClusterNotReady"
+	CompletedToClusterNotReady  Trigger = "Completed->ClusterNotReady"
+	OutOfDateToClusterNotReady  Trigger = "OutOfDate->ClusterNotReady"
+
+	// Transitions to InProgress state
+	ClusterNotReadyToInProgress Trigger = "ClusterNotReady->InProgress"
+	OutOfDateToInProgress       Trigger = "OutOfDate->InProgress"
+	CompletedToInProgress       Trigger = "Completed->InProgress"
+
+	// Transitions to TimedOut state
+	ClusterNotReadyToTimedOut Trigger = "ClusterNotReady->TimedOut"
+	InProgressToTimedOut      Trigger = "InProgress->TimedOut"
+
+	// Transitions to Completed state
```

Comment:
These transitions to `Completed` seem suspect to me.  Is it really possible to go directly from a `TimedOut` or `OutOfDate` state to `Completed` without doing any other work?

Reply:
Fixed, removed these transitions.
The following test had a tranisition from Timedout to Completed by mistake I believe, because of a missing timeout set to zero, so were essentially testing this condition:
- "Verify status conditions when configuration change causes ClusterInstance rendering to fail but configuration policy becomes compliant"
- "Verify status conditions when ClusterInstance provision has completed, ManagedCluster becomes ready and configuration policy becomes compliant"
This test tested it directly : Moves from TimedOut to Completed if all the policies are compliant. But that might be because of the previous 2?

Reply:
We can actually go from `TimedOut` to `Completed` if the policies become `Compliant` eventually and no other action is being taken as a consequence of the time out (which is the case right now).
It's also possible to go from `OutOfDate` to `Completed` if the `inform`, `NonCompliant` policy/policies that's causing the `OutOfDate` gets deleted and all the remaining policies are `Compliant`.

Reply:
Fixed with the latest updates. Redundant direct states removed. The machine each's state make their own decisions and transition between states until a final state has been reached.